core/src/test/java/org/apache/iceberg/hadoop/HadoopStreamsTest.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.hadoop;
+
+import java.io.IOException;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.s3a.S3ABlockOutputStream;
+import org.apache.iceberg.io.PositionOutputStream;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+class HadoopStreamsTest {
+
+  @Test
+  void closeShouldThrowIOExceptionOnInterrupted() throws IOException {
+
+    FSDataOutputStream fsDataOutputStream =
+        new FSDataOutputStream(new S3ABlockOutputStream(), null);
+    PositionOutputStream wrap = HadoopStreams.wrap(fsDataOutputStream);
+
+    // mock interrupt in S3ABlockOutputStream#putObject
+    Thread.currentThread().interrupt();
```

Comment:
It does check the scenario, if I remove the fix this test fails. This works as the interrupt is set in the same thread in https://github.com/apache/hadoop/blob/0f51d2a4ec17bad754beb17048409811a151be53/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3ABlockOutputStream.java#L635 as well

Reply:
I understood that test fails if fix is removed. but the scenario is not exactly the same, interrupt typically wasn't interrupted from the same thread. usually another thread is trying to interrupt a thread.

Reply:
yes but do we have to mock another thread interrupting current thread here if we are able to set the interrupt and mimic the behavior. 

I can start another thread in the test [S3ABlockOutputStream.java]'s close method and then interrupt it to set the interrupted flag but I feel that is not needed but nice to have as eventually all this will do is do what line #38 here is doing

Reply:
I would prefer that we change the test to interrupt the close from another thread. the close can sleep/block for 10 mins. interrupt should cancel the the close action and the assertion is that close takes less than 10 mins

Reply:
done